[13.0][MIG] account

MiquelRForgeFlow opened the pull request:
Migration of `account`.




--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

MiquelRForgeFlow committed:
[MIG] account

kos94ok-3D commented on addons/account/migrations/13.0.1.1/pre-migration.py:
```diff
@@ -0,0 +1,244 @@
+                })
+
+
+def create_account_move_new_columns(env):
+    """Faster way, avoid compute"""
+    openupgrade.logged_query(
+        env.cr, """
+        ALTER TABLE account_move
+        ADD COLUMN amount_total_signed numeric""",
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        ALTER TABLE account_move
+        ADD COLUMN amount_untaxed numeric""",
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        ALTER TABLE account_move
+        ADD COLUMN amount_untaxed_signed numeric""",
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        ALTER TABLE account_move
+        ADD COLUMN amount_tax numeric""",
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        ALTER TABLE account_move
+        ADD COLUMN amount_tax_signed numeric""",
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        ALTER TABLE account_move
+        ADD COLUMN amount_residual_signed numeric""",
+    )
```
Maybe rewrite like:
```
    data = {
        'account_move': [
            ('amount_total_signed', 'numeric'),
            ('amount_untaxed', 'numeric'),
            ('amount_untaxed_signed', 'numeric'),
            ('amount_tax', 'numeric'),
            ('amount_tax_signed', 'numeric'),
            ('amount_residual_signed', 'numeric'),
        ],
    }
    for table, column_spec_list in data.items():
        for column, column_type in column_spec_list:
            openupgrade.logged_query(
                env.cr, """
                ALTER TABLE {table}
                ADD COLUMN {column} {column_type}""".format(
                    table=table, column=column, column_type=column_type
                ),
            )
```
OR:
```
    data = {
        'account_move': [
            ('amount_total_signed', 'numeric'),
            ('amount_untaxed', 'numeric'),
            ('amount_untaxed_signed', 'numeric'),
            ('amount_tax', 'numeric'),
            ('amount_tax_signed', 'numeric'),
            ('amount_residual_signed', 'numeric'),
        ],
    }
    for table, column_spec_list in data.items():
        add_columns_command =  '\n'.join([
            'ADD COLUMN {column} {column_type}'
            for column, column_type in column_spec_list
        ])
        openupgrade.logged_query(
            env.cr,"""ALTER TABLE {table} \n{add_columns_command}""".format(
                table=table, add_columns_command=add_columns_command
            ),
        )
```

gdgellatly commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,719 @@
+        WHERE ai.state in ('draft', 'cancel')
+        RETURNING old_invoice_id, id""",
+    )
+    ids = [(x[0], x[1]) for x in env.cr.fetchall()]
```
```suggestion
    ids = env.cr.fetchall()
```
No point converting a list of tuple to a list of tuples.

gdgellatly commented on addons/account/migrations/13.0.1.1/pre-migration.py:
```diff
@@ -0,0 +1,244 @@
+        env.cr, """
+        ALTER TABLE account_move
+        ADD COLUMN old_invoice_id integer""",
+    )
```
I would also consider creating an index here. This column is used in a few queries as the join condition and will usually be a relatively large table.

EDIT: Maybe after the initial fill query below.

MiquelRForgeFlow replied:
will check

MiquelRForgeFlow replied:
done

gdgellatly commented:
ex tax invoices are showing the tax as an invoice line on historic invoices.

MiquelRForgeFlow commented:
> ex tax invoices are showing the tax as an invoice line on historic invoices.

yeah, and that's wrong? how should be handled correctly from your POV?

fmdl commented:
@MiquelRForgeFlow can make a litlle text to explain the methodoly to merge account.move and account.invoice ? Thanks

gdgellatly commented:
> > ex tax invoices are showing the tax as an invoice line on historic invoices.
> 
> yeah, and that's wrong? how should be handled correctly from your POV?

@MiquelRForgeFlow Sorry, I opened an issue with more detail that got closed before I added that 1 liner, which was that for this particular migration I could live with it.. Anyway, whether it is wrong or not I don't know. It certainly looks wrong, but I haven't done an actual invoice on v13 yet to know what looks right, but I imagine it should look the same as doing an invoice in v13.

khaeusler commented:
I've tried to migrate a v12 demo DB with the PR. In the demo DB, I added taxes to a product, created a sales order and then created an invoice.

The result on the migrated invoice is:
Untaxed Amount:	$ 14,750.00
Tax 15%:	$ 0.00
Total:	$ 16,962.50

Invoice in v12
![v12](https://user-images.githubusercontent.com/17731016/81866142-108ff600-956f-11ea-9a8b-a947d725344d.png)

Migrated invoice in v13
![v13](https://user-images.githubusercontent.com/17731016/81866159-15ed4080-956f-11ea-96f8-a69ee531f389.png)

If the Quantity is changed, the "Tax" in the footer changes correctly, but the "Untaced amount" is incorrectly changed to 0.
![v13_2](https://user-images.githubusercontent.com/17731016/81867534-23a3c580-9571-11ea-9807-507345fbcc1c.png)

sbidoul commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,719 @@
+        amount_tax, amount_total, state, invoice_payment_state, invoice_date,
+        invoice_date_due, invoice_payment_ref, old_voucher_id,
+        create_uid, create_date, write_uid, write_date)
+        SELECT message_main_attachment_id, number, date, reference, narration,
```
```suggestion
        SELECT message_main_attachment_id, COAELESCE(number, '/'), date, reference, narration,
```

Sometimes draft voucher have no number

khaeusler replied:
> Sometimes draft voucher have no number

After the migration, we still had invoices with name='' (empty value not null).

sbidoul commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,719 @@
+        FROM account_invoice_line ail
+        LEFT JOIN account_invoice ai ON ail.invoice_id = ai.id
+        LEFT JOIN account_move am ON am.old_invoice_id = ai.id
+        WHERE COALESCE(ai.move_id, am.id) IS NOT NULL""",
```
@MiquelRForgeFlow after migration I'm seeing duplicates move lines in validated invoices.
Don't we need to filter on `ai.state in ('draft', 'cancel')` here too?

sbidoul replied:
Also, this seems to create move lines with NULL debit and credit.

MiquelRForgeFlow replied:
The problem is that we cannot merge old move lines with old invoice lines, AFAIK, because there is not direct correspondence. Thus, at this moment what I do is mark old moves lines with "exclude_from_invoice_tab = TRUE" and transform old invoice lines to new move lines with "exclude_from_invoice_tab = False". Not sure if correct or if I should do something more/else. I am open to suggestions.

sbidoul replied:
For non draft invoices (ie invoices that have a move_id) a possibility could be to set `exclude_from_invoice_tab=TRUE` on move lines that are 
* on the payable/receivable account (which we know from the account field the invoice)
* or have `tax_line_id IS NOT NULL`

And for draft invoices I've not tested yet. Maybe it would be safer to let Odoo compute the move lines?

MiquelRForgeFlow replied:
and what about the move lines tat don't have those conditions? what about invoice lines?

sbidoul replied:
Normally these other move lines should be identical to invoice lines. Except when invoice line grouping was applied... in which case I don't see a solution to preserve the invoice lines.

MiquelRForgeFlow replied:
ok, can you check again a new migration?

sbidoul replied:
I will :)

sbidoul replied:
I'm still getting duplicate move lines on invoices (visible in the account move tab) with NULL debit and credit. I see the difficulty with merging invoice lines and move lines, and I don't see an easy solution for that, but I think that will cause troubles.

MiquelRForgeFlow commented:
@khaeusler thanks for your input. I will try to solve your case when having some time.

pedrobaeza commented:
Commit ae60003c91c9d524404d6de7ee074986f1e2cb0e shouldn't be here.

sbidoul commented:
@MiquelRForgeFlow I think there is an issue with the migration of invoice attachments.

MiquelRForgeFlow commented:
@sbidoul solved attachment issue

khaeusler commented:
> @khaeusler thanks for your input. I will try to solve your case when having some time.

OK 

Attachments are now available, but the taxes (value under the invoice lines) are still missing.

khaeusler commented:
Hello, the problem with the 0-tax occurs with draft bills, the field tax_base_amount of account.move.line has always the value 0.0

https://github.com/odoo/odoo/blob/13.0/addons/account/models/account_move.py#L1349

khaeusler commented:
I've corrected the bills as follows:
comment out the query for creating the useless tax lines
```diff
diff --git a/addons/account/migrations/13.0.1.1/post-migration.py b/addons/account/migrations/13.0.1.1/post-migration.py
index 0c415985eeb..9bb829f53c6 100644
--- a/addons/account/migrations/13.0.1.1/post-migration.py
+++ b/addons/account/migrations/13.0.1.1/post-migration.py
@@ -235,21 +235,21 @@ def migration_invoice_moves(env):
         WHERE COALESCE(ai.move_id, am.id) IS NOT NULL AND
             aa.internal_type in ('receivable', 'payable')""",
     )
-    openupgrade.logged_query(
-        env.cr, """
-        INSERT INTO account_move_line (company_id, account_id,
-        sequence, name, price_unit, currency_id, tax_base_amount,
-        tax_line_id, analytic_account_id, move_id, old_invoice_tax_id,
-        create_uid, create_date, write_uid, write_date)
-        SELECT ait.company_id, ait.account_id, ait.sequence, ait.name,
-        ait.amount, ait.currency_id, ait.base, ait.tax_id,
-        ait.account_analytic_id, COALESCE(ai.move_id, am.id),
-        ait.id, ait.create_uid, ait.create_date, ait.write_uid, ait.write_date
-        FROM account_invoice_tax ait
-        LEFT JOIN account_invoice ai ON ait.invoice_id = ai.id
-        LEFT JOIN account_move am ON am.old_invoice_id = ai.id
-        WHERE COALESCE(ai.move_id, am.id) IS NOT NULL""",
-    )
+    #openupgrade.logged_query(
+        #env.cr, """
+        #INSERT INTO account_move_line (company_id, account_id,
+        #sequence, name, price_unit, currency_id, tax_base_amount,
+        #tax_line_id, analytic_account_id, move_id, old_invoice_tax_id,
+        #create_uid, create_date, write_uid, write_date)
+        #SELECT ait.company_id, ait.account_id, ait.sequence, ait.name,
+        #ait.amount, ait.currency_id, ait.base, ait.tax_id,
+        #ait.account_analytic_id, COALESCE(ai.move_id, am.id),
+        #ait.id, ait.create_uid, ait.create_date, ait.write_uid, ait.write_date
+        #FROM account_invoice_tax ait
+        #LEFT JOIN account_invoice ai ON ait.invoice_id = ai.id
+        #LEFT JOIN account_move am ON am.old_invoice_id = ai.id
+        #WHERE COALESCE(ai.move_id, am.id) IS NOT NULL""",
+    #)
     openupgrade.logged_query(
         env.cr, """
         UPDATE account_invoice_payment_rel aipr
```

and then run these sql statements

```sql
-- Fix the name of invoices, some invoices had '' as number
UPDATE account_move SET name = '/' WHERE name='' AND state = 'draft';

-- Draft and Cancel invoices: fix values on account_move_line
UPDATE account_move_line aml
SET amount_residual=s.amount_residual,
    journal_id=s.journal_id,
    amount_residual_currency=s.amount_residual_currency,
    debit=s.debit,
    credit=s.credit,
    date=s.date,
    company_currency_id=s.company_currency_id,
    currency_id=s.currency_id,
    amount_currency=s.amount_currency,
    balance=s.balance,
    tax_exigible=s.tax_exigible,
    tax_base_amount=s.tax_base_amount,
    account_root_id=s.account_root_id,
    move_name = s.move_name,
    exclude_from_invoice_tab = False
FROM(
    SELECT
        aml.id,
        am.id AS move_id,
        am.type AS move_type,
        -ail.price_subtotal_signed AS amount_residual,
        am.journal_id AS journal_id,
        am.type,
        aml.price_subtotal,
        ail.price_subtotal_signed,
        CASE WHEN COALESCE(aml.currency_id, rc.currency_id)=rc.currency_id THEN 0.0
            ELSE -ail.price_total
        END AS amount_residual_currency,
        CASE WHEN am.type='out_invoice' THEN 0.0
            WHEN am.type='in_refund' THEN 0.0
            WHEN am.type='out_refund' THEN ABS(ail.price_subtotal_signed)
            WHEN am.type='in_invoice' THEN ABS(ail.price_subtotal_signed)
        END AS debit,
        CASE WHEN am.type='out_invoice' THEN ABS(ail.price_subtotal_signed)
            WHEN am.type='in_refund' THEN ABS(ail.price_subtotal_signed)
            WHEN am.type='out_refund' THEN 0.0
            WHEN am.type='in_invoice' THEN 0.0
        END AS credit,
        am.date AS date,
        rc.currency_id AS company_currency_id,
        NULLIF(aml.currency_id, rc.currency_id) AS currency_id,
        CASE WHEN COALESCE(aml.currency_id, rc.currency_id)=rc.currency_id THEN 0.0
            ELSE -aml.price_subtotal
        END AS amount_currency,
        -ail.price_subtotal_signed AS balance,
        TRUE AS tax_exigible,
        0 AS tax_base_amount,
        aa.root_id AS account_root_id,
        am.name AS move_name
    FROM account_move_line aml
    INNER JOIN account_invoice_line ail ON ail.id = aml.old_invoice_line_id
    INNER JOIN account_move am ON am.id = aml.move_id
    INNER JOIN res_company rc ON rc.id = aml.company_id
    INNER JOIN account_account aa ON aa.id = aml.account_id
    WHERE am.state IN ('draft', 'cancel') AND am.type != 'entry'
    ORDER BY am.id, aml.id
) AS s
WHERE aml.id=s.id;

-- Draft and Cancel invoices: Insert the tax lines to the journal items
INSERT INTO account_move_line (
    move_id,
    tax_line_id,
    create_date,
    account_id,
    company_id,
    currency_id,
    partner_id,
    create_uid,
    amount_residual,
    journal_id,
    amount_residual_currency,
    debit,
    credit,
    write_date,
    date,
    write_uid,
    name,
    company_currency_id,
    amount_currency,
    balance,
    quantity,
    tax_exigible,
    tax_base_amount,
    parent_state,
    account_internal_type,
    move_name,
    account_root_id,
    sequence,
    price_unit,
    discount,
    price_subtotal,
    price_total,
    tax_group_id,
    tax_repartition_line_id,
    is_rounding_line,
    exclude_from_invoice_tab
)
SELECT
    t.move_id,
    t.tax_line_id,
    t.create_date,
    atrl.account_id,
    t.company_id,
    t.currency_id,
    t.partner_id,
    t.create_uid,
    t.amount_residual * at.amount / -100 AS amount_residual,
    t.journal_id,
    t.amount_residual_currency,
    t.debit * at.amount / 100 AS debit,
    t.credit * at.amount / 100 AS credit,
    t.write_date,
    am.date,
    t.write_uid,
    at.name,
    t.company_currency_id,
    t.amount_currency * at.amount / -100 AS amount_currency,
    t.balance * at.amount / -100 AS balance,
    1.0 AS quantity,
    True AS tax_exigible,
    t.tax_base_amount,
    am.state AS parent_state,
    aa.internal_type AS account_internal_type,
    am.name AS move_name,
    aa.root_id AS account_root_id,
    10 AS sequence,
    t.price_unit * at.amount / 100 AS price_unit,
    t.discount,
    t.price_subtotal * at.amount / 100 AS price_subtotal,
    t.price_total * at.amount / 100 AS price_total,
    at.tax_group_id,
    atrl.id AS tax_repartition_line_id,
    False AS is_rounding_line,
    True AS exclude_from_invoice_tab
FROM (
    SELECT
        aml.move_id,
        amlatr.account_tax_id AS tax_line_id,
        MAX(aml.create_date) AS create_date,
        MAX(aml.company_id) AS company_id,
        MAX(NULLIF(aml.currency_id, rc.currency_id)) AS currency_id,
        MAX(aml.partner_id) AS partner_id,
        MAX(aml.create_uid) AS create_uid,
        SUM(ail.price_subtotal_signed) AS amount_residual,
        MAX(aml.journal_id) AS journal_id,
        CASE WHEN COALESCE(aml.currency_id, rc.currency_id)=rc.currency_id THEN 0.0
            ELSE SUM(-ail.price_total)
        END AS amount_residual_currency,
        CASE WHEN am.type='out_invoice' THEN 0.0
            WHEN am.type='in_refund' THEN 0.0
            WHEN am.type='out_refund' THEN SUM(ABS(ail.price_subtotal_signed))
            WHEN am.type='in_invoice' THEN SUM(ABS(ail.price_subtotal_signed))
        END AS debit,
        CASE WHEN am.type='out_invoice' THEN SUM(ABS(ail.price_subtotal_signed))
            WHEN am.type='in_refund' THEN SUM(ABS(ail.price_subtotal_signed))
            WHEN am.type='out_refund' THEN 0.0
            WHEN am.type='in_invoice' THEN 0.0
        END AS credit,
        MAX(aml.write_date) AS write_date,
        MAX(aml.write_uid) AS write_uid,
        MAX(aml.company_currency_id) AS company_currency_id,
        CASE WHEN COALESCE(aml.currency_id, rc.currency_id)=rc.currency_id THEN 0.0
            ELSE SUM(ail.price_subtotal_signed)
        END AS amount_currency,
        SUM(ail.price_subtotal_signed) AS balance,
        SUM(ail.price_subtotal_signed) AS tax_base_amount,
        MAX(aml.account_root_id) AS account_root_id,
        SUM(ail.price_subtotal_signed) AS price_unit,
        0.0 AS discount,
        SUM(ail.price_subtotal_signed) AS price_subtotal,
        SUM(ail.price_subtotal_signed) AS price_total,
        SUM(ail.price_subtotal_signed) AS price_subtotal_signed
    FROM account_move_line_account_tax_rel amlatr
    INNER JOIN account_move_line aml ON aml.id = amlatr.account_move_line_id
    INNER JOIN account_invoice_line ail ON ail.id = aml.old_invoice_line_id
    INNER JOIN res_company rc ON rc.id = aml.company_id
    INNER JOIN account_move am ON am.id = aml.move_id
    GROUP BY aml.move_id, amlatr.account_tax_id, aml.currency_id,rc.currency_id, am.type
) AS t
INNER JOIN account_move am ON am.id = t.move_id
INNER JOIN account_tax_repartition_line atrl ON atrl.repartition_type = 'tax'
INNER JOIN account_tax at ON at.id=t.tax_line_id AND at.amount != 0
INNER JOIN account_account aa ON aa.id = atrl.account_id
WHERE
    am.state IN ('draft', 'cancel')
    AND am.type != 'entry'
    AND (
        (am.type IN ('out_invoice', 'in_invoice') AND atrl.invoice_tax_id = t.tax_line_id)
        OR (am.type IN ('out_refund', 'in_refund') AND atrl.refund_tax_id = t.tax_line_id)
    )
ORDER BY t.move_id;

-- Draft and Cancel invoices: Insert the partner account to the journal items
INSERT INTO account_move_line (
    move_id,
    create_date,
    account_id,
    company_id,
    currency_id,
    partner_id,
    create_uid,
    amount_residual,
    journal_id,
    amount_residual_currency,
    debit,
    credit,
    write_date,
    date,
    write_uid,
    company_currency_id,
    amount_currency,
    balance,
    quantity,
    tax_exigible,
    tax_base_amount,
    parent_state,
    account_internal_type,
    move_name,
    account_root_id,
    sequence,
    price_unit,
    discount,
    price_subtotal,
    price_total,
    is_rounding_line,
    exclude_from_invoice_tab
)
SELECT
    t.move_id,
    am.create_date,
    SPLIT_PART(COALESCE(ip1.value_reference, ip2.value_reference), ',', 2)::INT AS account_id,
    am.company_id,
    t.currency_id,
    am.partner_id,
    am.create_uid,
    t.amount_residual,
    am.journal_id,
    t.amount_residual_currency,
    t.debit,
    t.credit,
    am.write_date,
    am.date,
    am.write_uid,
    t.company_currency_id,
    t.amount_currency,
    t.balance,
    1.0 AS quantity,
    True AS tax_exigible,
    0.0 AS tax_base_amount,
    am.state AS parent_state,
    aa.internal_type AS account_internal_type,
    am.name AS move_name,
    aa.root_id AS account_root_id,
    10 AS sequence,
    t.price_unit,
    0.0 AS discount,
    t.price_subtotal,
    t.price_total,
    False AS is_rounding_line,
    True AS exclude_from_invoice_tab
FROM(
    SELECT
    aml.move_id,
    MAX(currency_id) AS currency_id,
    MAX(company_currency_id) AS company_currency_id,
    -SUM(aml.amount_currency) AS amount_currency,
    -SUM(aml.amount_residual) AS amount_residual,
    -SUM(aml.amount_residual_currency) AS amount_residual_currency,
    SUM(aml.credit) AS debit,
    SUM(aml.debit) AS credit,
    -SUM(aml.balance) AS balance,
    -SUM(price_subtotal) AS price_unit,
    -SUM(price_subtotal) AS price_subtotal,
    -SUM(price_total) AS price_total
    FROM account_move_line aml
    GROUP BY aml.move_id
) AS t
INNER JOIN account_move am ON am.id = t.move_id AND am.state IN ('draft', 'cancel') AND am.type != 'entry'
LEFT JOIN ir_property ip1 ON ip1.company_id = am.company_id AND ip1.name='property_account_receivable_id' AND ip1.res_id='res.partner,' || am.partner_id
LEFT JOIN ir_property ip2 ON ip2.company_id = am.company_id AND ip2.name='property_account_receivable_id' AND ip2.res_id IS NULL
INNER JOIN account_account aa ON aa.id = SPLIT_PART(COALESCE(ip1.value_reference, ip2.value_reference), ',', 2)::INT;

-- Posted invoices: Fix values on account_move_line
UPDATE account_move_line aml
SET
    date_maturity=s.date_maturity,
    openupgrade_legacy_13_0_user_type_id=s.openupgrade_legacy_13_0_user_type_id,
    blocked=s.blocked,
    analytic_account_id=s.analytic_account_id,
    create_uid=s.create_uid,
    amount_residual=s.amount_residual,
    journal_id=s.journal_id,
    amount_residual_currency=s.amount_residual_currency,
    debit=s.debit,
    debit_cash_basis=s.debit_cash_basis,
    reconciled=s.reconciled,
    credit=s.credit,
    balance_cash_basis=s.balance_cash_basis,
    write_date=s.write_date,
    date=s.date,
    write_uid=s.write_uid,
    move_id=s.move_id,
    company_currency_id=s.company_currency_id,
    currency_id=NULLIF(aml.currency_id, s.company_currency_id),
    product_id=s.product_id,
    invoice_id=s.invoice_id,
    credit_cash_basis=s.credit_cash_basis,
    product_uom_id=s.product_uom_id,
    amount_currency=s.amount_currency,
    balance=s.balance,
    quantity=s.quantity,
    tax_exigible=s.tax_exigible,
    tax_base_amount=s.tax_base_amount,
    parent_state=s.parent_state,
    account_internal_type=s.account_internal_type,
    move_name=s.move_name,
    account_root_id=s.account_root_id,
    exclude_from_invoice_tab = False
FROM(
    SELECT
        aml.date_maturity,
        aml.openupgrade_legacy_13_0_user_type_id,
        aml.blocked,
        aml.analytic_account_id,
        aml.create_uid,
        aml.amount_residual,
        aml.journal_id,
        aml.amount_residual_currency,
        aml.debit,
        aml.debit_cash_basis,
        aml.reconciled,
        aml.credit,
        aml.balance_cash_basis,
        aml.write_date,
        aml.date,
        aml.write_uid,
        aml.move_id,
        aml.company_currency_id,
        aml.product_id,
        aml.invoice_id,
        aml.credit_cash_basis,
        aml.product_uom_id,
        aml.amount_currency,
        aml.balance,
        aml.quantity,
        aml.tax_exigible,
        aml.tax_base_amount,
        aml.parent_state,
        aml.account_internal_type,
        aml.move_name,
        aml.account_root_id
    FROM account_move_line aml
    INNER JOIN account_move am ON am.id = aml.move_id AND am.state='posted' AND am.type != 'entry'
    WHERE aml.product_id IS NOT NULL AND aml.exclude_from_invoice_tab IS TRUE
) AS s
WHERE
    aml.move_id = s.move_id
    AND aml.product_id = s.product_id
    AND aml.quantity = s.quantity
    AND (
        ABS(aml.price_subtotal) = ABS(s.amount_currency)
        OR ABS(aml.price_subtotal) = ABS(s.balance)
    )
    AND aml.exclude_from_invoice_tab IS NOT TRUE;

-- Posted invoices: Delete duplicate lines on account_move_line
DELETE FROM account_move_line WHERE id IN (
    SELECT aml.id
    FROM account_move_line aml
    INNER JOIN account_move am ON am.id = aml.move_id AND am.state='posted' AND am.type != 'entry'
    WHERE aml.product_id IS NOT NULL AND aml.exclude_from_invoice_tab IS TRUE
);

-- Determine and set tax_repartition_line_id
UPDATE account_move_line aml
SET tax_repartition_line_id=s.tax_repartition_line_id
FROM(
    SELECT
        aml.id,
        aml.move_id,
        atrl.id AS tax_repartition_line_id
    FROM account_move_line aml
    INNER JOIN account_move am ON am.id = aml.move_id AND am.type IN ('out_invoice', 'in_invoice', 'out_refund', 'in_refund')
    INNER JOIN account_tax_repartition_line atrl ON atrl.repartition_type='tax' AND atrl.account_id = aml.account_id
    WHERE
        aml.tax_line_id IS NOT NULL
        AND tax_repartition_line_id IS NULL
        AND (
            (am.type IN ('out_invoice', 'in_invoice') AND atrl.invoice_tax_id = aml.tax_line_id)
            OR (am.type IN ('out_refund', 'in_refund') AND atrl.refund_tax_id = aml.tax_line_id)
        )

) s
WHERE aml.id = s.id;

-- invoices: Set exclude_from_invoice_tab to True on invoice lines
UPDATE account_move_line aml
SET
    exclude_from_invoice_tab = False
FROM(
    SELECT
        aml.id
    FROM account_move_line aml
    INNER JOIN account_move am ON am.id = aml.move_id AND am.type != 'entry'
    WHERE aml.old_invoice_line_id IS NOT NULL AND aml.exclude_from_invoice_tab IS NULL
) AS s
WHERE
    aml.id = s.id;

-- Set the correct amount on 2 columns in account_move
UPDATE account_move am
SET
    amount_total_signed = ai.amount_total_company_signed,
    amount_residual_signed = ai.residual_company_signed
FROM account_invoice ai
WHERE 
    ai.id = am.old_invoice_id;
```

left before and right after my fixes:
![corrected_invoices](https://user-images.githubusercontent.com/17731016/85211296-0d471180-b348-11ea-944d-8574fb825203.png)

EDIT: minor adjustments on the SQL statements so that the reporting works
EDIT2: fix for company amounts and set the exclude_from_invoice_tab

odix67 commented:
> I've corrected the bills as follows:
> comment out the query for creating the useless tax lines
> 
> ```diff
> diff --git a/addons/account/migrations/13.0.1.1/post-migration.py b/addons/account/migrations/13.0.1.1/post-migration.py
> index 0c415985eeb..9bb829f53c6 100644
> --- a/addons/account/migrations/13.0.1.1/post-migration.py
> +++ b/addons/account/migrations/13.0.1.1/post-migration.py

I'm a little bit confused not finding the mentioned post-migration.py file, or do I just sitting on my eyes ;-)

ok, found, was sitting on my eyes, sorry

MiquelRForgeFlow commented:
@khaeusler I think I have fixed your issue in https://github.com/OCA/OpenUpgrade/pull/2275#issuecomment-628255038.

About the move lines thing, do you mind if I do a verbatim copy of https://github.com/OCA/OpenUpgrade/pull/2275#issuecomment-647043843? Not tested, but looks good.

pedrobaeza commented:
@MiquelRForgeFlow I'm now analyzing this addon, so please keep it unaltered and I'll check the SQL and incorporate it if needed.

khaeusler commented:
Hello,

@MiquelRForgeFlow yes, you can use the sql statement; that's why I posted it.

It would be good if someone else tested the SQL statements.
It looks OK in _our_ customer database.

kos94ok-3D commented:
Hi, @pedrobaeza
Any news about a module?

pedrobaeza commented:
I'm checking it.

kos94ok-3D commented:
Hello, @pedrobaeza. How can I help you?
Maybe do you haven`t time to resolve some migration problems?
You can write to me, if need.

pedrobaeza commented:
Yes, I haven't been able yet to complete it, as being on holidays half of my team, customer issues avoids me to work on this. I'm evolving the code for not duplicating move lines for invoices and for journal items.

kos94ok-3D committed:
[FIX] account: don't duplicate aml for invoice with state open or paid

kos94ok-3D commented:
@pedrobaeza, I didn't do for draft invoices but for open and paid invoices work this code [my improvements with minimum diff](https://github.com/kos94ok-3D/OpenUpgrade/compare/13.0-mig-account-script...kos94ok-3D:my_accounting_fix)
Check it please, when you have time.

kos94ok-3D commented:
@pedrobaeza added fixes for draft invoices link the same as above [link with diff](https://github.com/kos94ok-3D/OpenUpgrade/compare/13.0-mig-account-script...kos94ok-3D:my_accounting_fix)

kos94ok-3D commented on addons/account/migrations/13.0.1.1/openupgrade_analysis_work.txt:
```diff
@@ -0,0 +1,822 @@
+# NOTHING TO DO: it's somewhat related to the removed 'multi' from 'ir.actions.act_window'
+
+account      / account.payment.term.line / option (selection)            : selection_keys is now '['day_after_invoice_date', 'day_current_month', 'day_following_month']' ('['after_invoice_month', 'day_after_invoice_date', 'day_current_month', 'day_following_month']')
+# DONE: post-migration: reallocated 'after_invoice_month' option to 'day_following_month'
```
Since https://github.com/odoo/odoo/pull/57537 it was revert

pedrobaeza committed:
[FIX+IMP] account: Fill default new accounts in company + don't require old_invoice in SQL

v13 introduces some new accounts at company level. We need to ensure
that the information is filled on existing companies for avoiding problems
in  subsequent modules (like `point_of_sale`).

etobella commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,735 @@
+        INSERT INTO account_move_line (company_id, account_id,
+        sequence, name, price_unit, currency_id, tax_base_amount,
+        tax_line_id, analytic_account_id, move_id, old_invoice_tax_id,
+        create_uid, create_date, write_uid, write_date)
```
it is needed to add exclude_from_invoice_tab as TRUE here, because it is showing the tax as a line:

![imagen](https://user-images.githubusercontent.com/28590170/94536839-18f97c80-0243-11eb-9eca-808c1f270d8f.png)

pedrobaeza replied:
This code is not definitive. I will push a later version today.

etobella replied:
Thanks @pedrobaeza I was just making a comment about an issue I found :smile: 

By the way, great work on OpenUpgrade!!!

pedrobaeza replied:
Please try now with the pushed version if you still have the same problem.

etobella reviewed (commented):
@JordiBForgeFlow @MiquelRForgeFlow I found a small issue

kos94ok-3D commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,779 @@
+    )
+
+
+def fill_account_journal_restrict_mode_hash_table(env):
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE account_journal
+        SET restrict_mode_hash_table = TRUE
+        WHERE {} != TRUE
+        """.format(openupgrade.get_legacy_name('update_posted'))
+    )
```
Need to fill secure_sequence_id.
As I remember, I couldn`t confirm the invoice drafted
```python
def fill_account_journal_restrict_mode_hash_table(env):
    openupgrade.logged_query(
        env.cr, """
        UPDATE account_journal
        SET restrict_mode_hash_table = TRUE
        WHERE {} != TRUE
        RETURNING id
        """.format(openupgrade.get_legacy_name('update_posted'))
    )
    ids = [x[0] for x in env.cr.fetchall()]
    if ids:
        env['account.journal'].browse(ids)._create_secure_sequence(['secure_sequence_id'])
```

kos94ok-3D commented on addons/account/migrations/13.0.1.1/pre-migration.py:
```diff
@@ -0,0 +1,272 @@
+    ('account.move', 'account_move', 'reverse_entry_id', 'reversed_entry_id'),
+]
+
+_field_sale_renames = [
+    ('res.company', 'res_company', 'sale_note', 'invoice_terms'),
```
<pre>
('res.config.settings', 'res_config_settings', 'use_sale_note', 'use_invoice_terms'),
</pre>

kos94ok-3D commented on addons/account/migrations/13.0.1.1/pre-migration.py:
```diff
@@ -0,0 +1,272 @@
+    openupgrade.copy_columns(cr, _column_copies)
+    openupgrade.rename_columns(cr, _column_renames)
+    openupgrade.rename_fields(env, _field_renames)
+    if openupgrade.table_exists(cr, 'sale_order'):
+        openupgrade.rename_fields(env, _field_sale_renames)
```
<pre>
# https://github.com/odoo/odoo/commit/ca25a692bd19fdca2b2600f2054eb419aae28999#diff-3bf7fa37ad269bad5d8df8214bd5206b
openupgrade.logged_query(env.cr, """
            UPDATE ir_config_parameter
            SET key = 'account.use_invoice_terms'
            WHERE key = 'sale.use_sale_note'
        """)
</pre>

kos94ok-3D reviewed (commented):
My additional fixes

kos94ok-3D commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,779 @@
+        ai.amount_untaxed_signed, ai.amount_tax_company_signed, COALESCE(
+        ai.amount_total_company_signed, ai.amount_total_signed),
+        COALESCE(ai.residual_company_signed, ai.residual_signed),
+        ai.fiscal_position_id, ai.user_id, CASE WHEN ai.state = 'cancel'
+        THEN 'cancel' ELSE 'posted' END, CASE WHEN ai.state IN (
```
Case statement not needed because records filtered by state
<pre>ai.fiscal_position_id, ai.user_id, 'posted', CASE WHEN ai.state IN (</pre>

kos94ok-3D commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,779 @@
+        COALESCE(av.date, am.date), av.reference, av.narration,
+        CASE WHEN av.voucher_type = 'purchase' THEN 'in_receipt'
+        ELSE 'out_receipt' END, av.journal_id, av.company_id,
+        av.currency_id, av.partner_id, av.tax_amount, av.amount,
+        CASE WHEN av.state = 'proforma' THEN 'posted' ELSE av.state END,
+        CASE WHEN av.state = 'cancel' THEN 'not_paid' ELSE 'paid' END,
```
Case statement not needed because records filtered by state
<pre>av.currency_id, av.partner_id, av.tax_amount, av.amount, 'posted', 'paid',</pre>

kos94ok-3D commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,779 @@
+        SELECT message_main_attachment_id, COALESCE(number, '/'), date,
+        reference, narration, CASE WHEN voucher_type = 'purchase'
+        THEN 'in_receipt' ELSE 'out_receipt' END, journal_id, company_id,
+        currency_id, partner_id, tax_amount, amount, state, 'not_paid',
```
fix for account voucher with `state == proforma`
<pre>
currency_id, partner_id, tax_amount, amount, CASE WHEN av.state = 'proforma'
THEN 'posted' ELSE av.state END, 'not_paid',
</pre>

kos94ok-3D commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,779 @@
+        JOIN account_account aa ON aaat.account_account_id = aa.id
+        JOIN res_company rc ON aa.company_id = rc.id
+        JOIN res_partner rp ON rc.partner_id = rp.id
+        WHERE aat.applicability = 'accounts' AND aat.country_id IS NULL
+            AND aaat.account_account_tag_id = aat.id
+            AND rp.country_id IS NOT NULL""",
```
I think we can remove it because we have this filter in `select_tags` table and inner join with this table

kos94ok-3D commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,779 @@
+        JOIN account_tax t ON atat.account_tax_id = t.id
+        JOIN res_company rc ON t.company_id = rc.id
+        JOIN res_partner rp ON rc.partner_id = rp.id
+        WHERE aat.applicability = 'taxes' AND aat.country_id IS NULL
+            AND atat.account_account_tag_id = aat.id
+            AND rp.country_id IS NOT NULL""",
```
I think we can remove it because we have this filter in `select_tags` table and inner join with this table

kos94ok-3D commented on addons/account/migrations/13.0.1.1/pre-migration.py:
```diff
@@ -0,0 +1,272 @@
+    index_name = '%s_%s_index' % ("account_move", "old_invoice_id")
+    sql.create_index(
+        env.cr, index_name, "account_move", ['"%s"' % "old_invoice_id"])
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE account_move am
+        SET old_invoice_id = aml.invoice_id
+        FROM account_move_line aml
+        WHERE aml.move_id = am.id AND am.old_invoice_id IS NULL
+        """,
+    )
+    openupgrade.logged_query(
+        env.cr, """
+        UPDATE account_invoice ai
+        SET move_id = am.old_invoice_id
+        FROM account_move am
+        WHERE am.old_invoice_id = ai.id AND ai.move_id IS NULL
+        """,
+    )
```
I think we can replace this just one query
<pre>
    openupgrade.logged_query(
        env.cr, """
        UPDATE account_invoice ai
        SET move_id = aml.move_id
        FROM account_move_line aml
        WHERE aml.invoice_id = ai.id AND ai.move_id IS NULL
        """,
    )
</pre>
but need write this before query:
<pre>
    openupgrade.logged_query(
        env.cr, """
        UPDATE account_move am
        SET old_invoice_id = ai.id
        FROM account_invoice ai
        WHERE ai.move_id = am.id
        """,
    )
</pre>

kos94ok-3D reviewed (commented):
Some performance improvements

etobella commented:
@pedrobaeza It's not working, but the error is different.

I created 2 journals, one grouped and the other one not. I created an invoice for each one with two lines, after the migration, the invoice from the journal without group had two lines, but the other one had been joined. I used no product (it was a demo database), I don't know if it can be an issue.

pedrobaeza commented:
@etobella I'm afraid that conditions can't be fixed easily, as grouped journal items don't have a correspondence with invoice lines, so both need to be put. I don't have any customer with that case, but if it's yours, it will require to analyze how users have been working with this and get to an heuristics, but not fixed method.

etobella commented:
Well, in my case, all the invoices we made have this issue :dizzy_face: 
Also, the original odoo module could be overridden to decide how to group the invoices, this could make it even harder to fix :sob: .

The previous method showed the correct data (before the last commits), the issue I found there was related to taxes, that could be solved with my comment.

pedrobaeza commented:
But previous code was duplicating lines for invoices and for journal items, and only tweaking the visibility through `exclude_invoice_tab` flag, but this will create other problems. I recommend you a custom migration scripts for dealing with your custom group way (or to use previous code), but I don't see the tax problem so easily to be solved as you say.

Maybe we can put both codes here with a conditional depending on the grouping flag on the journal. With this, I think both will be satisfied.

MiquelRForgeFlow commented:
> Maybe we can put both codes here with a conditional depending on the grouping flag on the journal. With this, I think both will be satisfied.

I was thinking the same haha

kos94ok-3D committed:
[FIX] Draft tax lines and other fixes

kos94ok-3D commented:
HI, @etobella
Try to do this change in script [link](https://github.com/kos94ok-3D/OpenUpgrade/compare/13.0-mig-account-script...fix_draft_tax_lines)

gdgellatly commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,781 @@
+    )
+    # Compute balance for Draft Invoice Lines
+    draft_invoices = env['account.move'].search([('state', 'in', ('draft', 'cancel'))])
+    for invoice_line in draft_invoices.line_ids:
```
I believe we might have a performance issue here. I have a client with not a massive number of draft vendor bills, say 3,500, and stuck here for 16 hours so far. As it happens, they are all errors anyway so I can just delete and restart but FYI.

UPDATE: Before deleting draft invoices - QUIT after 16 hours, after deletion entire upgrade run in 20 mins.

pedrobaeza replied:
OK, thanks for the info. I will check for alternatives.

kos94ok-3D replied:
@gdgellatly try to use with this changes:
<pre>
    draft_invoices = env['account.move'].search([('state', 'in', ('draft', 'cancel'))])
    for invoice_line in draft_invoices.line_ids:
        vals = invoice_line._get_fields_onchange_subtotal()
        invoice_line.write(vals)
    draft_invoices.with_context(check_move_validity=False)._recompute_dynamic_lines()
</pre>
<pre>
    draft_invoices = env['account.move'].search([('state', 'in', ('draft', 'cancel'))]).with_context(check_move_validity=False)
    draft_invoices.line_ids.read([
        'price_unit', 'quantity', 'discount', 'currency_id', 'product_id',
        'partner_id', 'tax_ids', 'move_id', 'price_subtotal', 'recompute_tax_line',
        'exclude_from_invoice_tab', 'amount_currency', 'balance',
        'tax_repartition_line_id',
    ])
    draft_invoices.line_ids._onchange_price_subtotal()
    draft_invoices._recompute_dynamic_lines(recompute_all_taxes=True)
</pre>

gdgellatly replied:
@kos94ok-3D As I said, in this case it was an error in client database, so I deleted them all now anyway.

MiquelRForgeFlow commented:
Everybody following this PR, please, could you test ASAP this PR to see what's left to do or needs to be fixed? I hope to finish it after your inputs!

kos94ok-3D commented on addons/account/migrations/13.0.1.1/post-migration.py:
```diff
@@ -0,0 +1,793 @@
+    fill_account_account_type_internal_group(env)
+    map_account_journal_post_at_bank_rec(env)
+    fill_account_journal_restrict_mode_hash_table(env)
+    map_account_payment_term_line_option_after_invoice_month(env)
```
Since odoo/odoo#57537 it was revert

MiquelRForgeFlow replied:
Yeah. I will update it when https://github.com/OCA/OpenUpgrade/pull/2409 gets merged.

kos94ok-3D replied:
OK, Thanks.

kos94ok-3D reviewed (approved):
I can`t test migration these two weeks, but I tested with this change earlier.

nomadic-coding commented:
Hi, thanks for the work on the account migration. I encountered three issues related to migrated **draft** invoices.
1. when posting: Required field missing Model: Analytic Line (account.analytic.line), Field: Date (date)
2. Invoice total is 0
3. When posting there is an error about unbalanced journal entry.

![Selection_221](https://user-images.githubusercontent.com/38637556/95689468-41a14f00-0c11-11eb-8903-3a9e651e8223.png)
![Selection_231](https://user-images.githubusercontent.com/38637556/95689772-5da5f000-0c13-11eb-8d6a-da54185a1bf6.png)

kos94ok-3D commented:
@nomadic-coding try to use with these changes: https://github.com/kos94ok-3D/OpenUpgrade/commit/d7748668a47ba933885941a40c6706b8a3ba3837

MiquelRForgeFlow committed:
[IMP] link move lines with account tags and fix case for grouped journal items

MiquelRForgeFlow commented:
Hello everybody following this PR, I have added the relation between move lines and requisition lines and account tags. Could you do one last review?

@etobella I have implemented the other way for grouped journal items. Could you check again?

@nomadic-coding could you check again if your issues persist?

nomadic-coding commented:
@MiquelRForgeFlow @kos94ok-3D  thanks for the fixes. Looks good now. Well done. I'll keep on testing.

etobella reviewed (approved):
Tested with aggregated invoices and not and all works as expected

pedrobaeza commented:
OK, let's merge this as a good base and we can improve later if we find more things. Thanks to all the involved people in developing and reviewing.

BartvandenHout commented:
I tried a v12 > v13 upgrade today. Seems all but account is working. I still don't get the correct invoices. On my v13-database the account_move and account_move_line I get, seems to me, incorrect or no values:

**account_move**
Not set correctly:
type = 'entry' on all records
Not set al all:
to_check, amount_untaxed, amount_tax, amount_residual, amount_untaxed_signed, amount_tax_signed, amount_residual_signed, fiscal_position_id, invoice_payment_state, invoice_date, invoice_date_due, invoice_payment_ref, invoice_sent, invoice_origin

**account_move_line**
Not set:
amount_residual, is_rounding_line, exclude_from_invoice_tab

Am I forgetting something? Is my v12-dbase not correct (seems to be, though)? Is it still work-in-progress?

FYI, to get the latest OpenUpgrade installation I installed today:
pip3 install git+git://github.com/OCA/openupgradelib.git
git clone https://github.com/OCA/OpenUpgrade.git
cd OpenUpgrade
git pull origin 13.0

Btw. Thanks for all the work you put into it

pedrobaeza commented:
Sorry, but everything is working in our side, so you should have something incorrect in your end, or some non described steps to get to that problem, so unless you bring more information to shed some light, nothing can be done.

BartvandenHout commented:
Thank for your reply. I tried my method with a new database, and no invoices either. So it's definitely on my side. Can you see what's wrong with my current process?

Deleted post partially: an incorrect method should not be found on the internet

Any advice would be greatly appreciated!

pedrobaeza commented:
It should fall on data side, so I'm afraid I can't help you

MiquelRForgeFlow commented:
@BartvandenHout maybe you will find a clue in the logs of your migration

BartvandenHout commented:
Deleted post: wrong place for posting my method. I'll find a better place for it :-)

luistriana28 commented:
Today I tried the script, all the migration was a success process.
Only detail is all the invoices are in "entry" views.

Any update about it?

pedrobaeza commented:
@luistriana28 this is working well in a lot of DBs, so you should have something strange in your side.

pedrobaeza commented:
Unless you mean that the invoices are seen both in invoices menu and journal entries menu, which is normal and the expected behavior.

luistriana28 commented:
@pedrobaeza So there is something in the script even with the query in DB
SELECT * FROM account_move where type='out_invoice';
are 0 queries.
![Captura de Pantalla 2021-09-29 a la(s) 12 13 12](https://user-images.githubusercontent.com/12242599/135317239-c6656448-f2f6-499b-a434-b7f7dab1f9bd.png)

So after checking the views they are a non modified. 

![Captura de Pantalla 2021-09-29 a la(s) 12 12 52](https://user-images.githubusercontent.com/12242599/135317291-4e6581cb-03f0-4037-99f5-cfc2e7b06496.png)

pedrobaeza commented:
Then I come back to my initial hypothesis: there's something incorrect in your original DB. Please refrain from writing here more, and if you have enough information for being considered a bug in the scripts, open an issue.
